.github/workflows/scripts/setup_bundler

```diff
@@ -18,26 +37,86 @@ if [[ $RAILS_VERSION =~ .*"4".* ]]; then
   og_bundler=$(bundle --version | awk '{ print $3 }')
   echo "DEBUG: pre-installed Bundler version is $og_bundler"
-  if [[ "$og_bundler" != "1.17.3" ]]; then
-    # need to use bundler <2 for all rails 4 and below tests
-    echo "DEBUG: running 'gem install bundler'"
-    gem install --default bundler:1.17.3 --force
-    echo "DEBUG: running 'gem list bundler'"
+  if [[ "$og_bundler" = "1.17.3" ]]; then
+    return
+  fi
+  # need to use bundler <2 for all rails 4 and below tests
+  echo "DEBUG: running 'gem install bundler'"
+  gem install --default bundler:1.17.3 --force
+  echo "DEBUG: running 'gem list bundler'"
+  gem list bundler
+  echo "DEBUG: RUBY_VERSION = $RUBY_VERSION"
+  if [[ $RUBY_VERSION != "2.2.10" ]]; then 
+    echo "DEBUG: running 'gem uninstall bundler'"
+    gem uninstall bundler
+    echo "DEBUG: running 'gem list bundler' again"
     gem list bundler
-    echo "DEBUG: RUBY_VERSION = $RUBY_VERSION"
-    if [[ $RUBY_VERSION != "2.2.10" ]]; then 
-      echo "DEBUG: running 'gem uninstall bundler'"
-      gem uninstall bundler
-      echo "DEBUG: running 'gem list bundler' again"
-      gem list bundler
-    fi
   fi
+}
+
+function configure_bundler {
+  if ! [[ $RUBY_VERSION =~ 2.[23] ]]; then
+    echo "DEBUG: Ruby is not at version 2.2-2.5, skipping 'bundler config'"
+    return
+  fi
+
   # add mysql specific config for bundler when we are using older mysql
   echo "DEBUG: running 'bundle config'"
   bundle config --global build.mysql2 --with-mysql-config=/usr/local/mysql55/bin/mysql_config
-fi
+}
+
+function install_ruby_version_specific_gems {
+  if using_jruby; then
+    echo "DEBUG: Skipping specific gem installation, as JRuby is in use"
+    return
+  fi
+
+  # for some reason, ruby 3.0 fails unless 3.1.0 is installed before it bundles for rails_prepend suite/rails 7
+  if [[ $RUBY_VERSION =~ 3.0 ]]; then 
+    echo "DEBUG: running 'gem install digest'"
+    gem install --default digest:3.1.0
+    echo "installed digest 3.1.0"
+  fi
+
+  # for some reason, ruby 2.7+ fails unless strscan 3.0.4 is installed
+  if [[ $RUBY_VERSION = 3.* || $RUBY_VERSION =~ 2.7 ]]; then 
+    echo "DEBUG: running 'gem install strscan'"
+    gem install --default strscan:3.0.4
+    echo "installed strscan 3.0.4"
+
+    echo "DEBUG: running gem install set"
+    gem install --default set:1.0.3
+    echo "installed set 1.0.3"
+  fi
+}
+
+function set_up_bundler {
+  if using_jruby; then
+    echo "DEBUG: Skipping Bundler setup, as JRuby is in use"
+    return
+  fi
+
+  if ! using_old_rails && ! using_old_ruby; then
+    echo "DEBUG: Skipping Bundler setup, as neither an old Rails or Ruby version is in use"
+    return
+  fi
+
+  update_to_desired_rubygems_version
+  install_desired_bundler_version
+  configure_bundler
+}
+
+echo "DEBUG: setting up Bundler"
+set_up_bundler
+
 echo "DEBUG: running 'bundle install'"
 bundle install
+
+echo "DEBUG: installing Ruby version specific gems"
+install_ruby_version_specific_gems
```

Comment:
This might need to be before the bundle install, but I can't remember when it failed before. I know it would fail on a bundle install, but I can't remember if it was the one in this file, or one in a multiverse suite.

Reply:
We could keep it like this though and see if any of those failures eventually pop up, and then if they do we could just move it

Reply:
(Discussed offline) we're going to leave things as-is for now and circle back if we see any related issues crop up in the future.